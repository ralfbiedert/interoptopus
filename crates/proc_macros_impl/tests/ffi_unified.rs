use quote::quote;
use syn::{parse_quote, ItemFn, ItemStruct, ItemEnum, ItemImpl};

mod util;

#[test]
fn ffi_struct_expansion() {
    let item: ItemStruct = parse_quote! {
        #[ffi]
        pub struct TestStruct {
            pub x: u32,
            pub y: u32,
        }
    };

    insta::assert_snapshot!(expand_ffi!(item));
}

#[test]
fn ffi_enum_expansion() {
    let item: ItemEnum = parse_quote! {
        #[ffi]
        pub enum TestEnum {
            Variant1,
            Variant2(u32),
        }
    };

    insta::assert_snapshot!(expand_ffi!(item));
}

#[test]
fn ffi_function_expansion() {
    let item: ItemFn = parse_quote! {
        #[ffi]
        pub fn test_function(input: u32) -> u32 {
            input * 2
        }
    };

    insta::assert_snapshot!(expand_ffi!(item));
}

#[test]
fn ffi_impl_expansion() {
    let item: ItemImpl = parse_quote! {
        #[ffi]
        impl TestService {
            pub fn new(value: u32) -> Self {
                Self { value }
            }

            pub fn get_value(&self) -> u32 {
                self.value
            }
        }
    };

    insta::assert_snapshot!(expand_ffi!(item));
}