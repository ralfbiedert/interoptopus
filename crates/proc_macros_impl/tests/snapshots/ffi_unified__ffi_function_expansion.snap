---
source: crates/proc_macros_impl/tests/ffi_unified.rs
expression: expand_ffi!(item)
---
#[unsafe(export_name = "test_function")]
pub extern "C" fn test_function(input: u32) -> u32 {
    input * 2
}
#[allow(non_camel_case_types)]
pub struct test_function {}
const _: () = const {
    ::interoptopus::lang::types::assert_raw_safe::<u32>();
};
const _: () = const {
    ::interoptopus::lang::types::assert_raw_safe::<u32>();
};
impl ::interoptopus::lang::function::FunctionInfo for test_function {
    fn id() -> ::interoptopus::inventory::FunctionId {
        ::interoptopus::inventory::FunctionId::from_id(
            ::interoptopus::id!(test_function),
        )
    }
    fn signature() -> ::interoptopus::lang::function::Signature {
        ::interoptopus::lang::function::Signature {
            arguments: vec![
                ::interoptopus::lang::function::Argument::new("input", < u32 as
                ::interoptopus::lang::types::TypeInfo > ::id())
            ],
            rval: <u32 as ::interoptopus::lang::types::TypeInfo>::id(),
        }
    }
    fn function() -> ::interoptopus::lang::function::Function {
        ::interoptopus::lang::function::Function {
            name: "test_function".to_string(),
            visibility: ::interoptopus::lang::meta::Visibility::Public,
            docs: ::interoptopus::lang::meta::Docs::from_lines(vec![]),
            emission: ::interoptopus::lang::meta::Emission::External,
            signature: Self::signature(),
        }
    }
    fn register(inventory: &mut ::interoptopus::inventory::Inventory) {
        <u32 as ::interoptopus::lang::types::TypeInfo>::register(inventory);
        <u32 as ::interoptopus::lang::types::TypeInfo>::register(inventory);
        inventory.register_function(Self::id(), Self::function());
    }
}
